### PR TITLE
port dotnet/corefx#11348

### PR DESCRIPTION
**Pull request by analogrelay**
also adds some tests and extra features to the EchoApp test sample

/cc @Tratcher @moozzyk

### CODE REVIEW

File: samples/EchoApp/Startup.cs [hunk 1]
```diff
@@ -49,27 +49,57 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerF
...
+        private async Task Echo(HttpContext context, WebSocket webSocket, ILogger logger)
         {
             var buffer = new byte[1024 * 4];
             var result = await webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);
+            LogFrame(logger, result, buffer);
             while (!result.CloseStatus.HasValue)
             {
                 // If the client send "ServerClose", then they want a server-originated close to occur
-                if(result.MessageType == WebSocketMessageType.Text)
+                string content = "<<binary>>";
```
Comment (moozzyk) [inline samples/EchoApp/Startup.cs]: nit: `var`

File: samples/EchoApp/Startup.cs [hunk 2]
```diff
@@ -49,27 +49,57 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerF
...
+        {
+            var close = frame.CloseStatus != null;
+            string message;
+            if (close)
+            {
+                message = $"Close: {frame.CloseStatus.Value} {frame.CloseStatusDescription}";
+            }
+            else
+            {
+                string content = "<<binary>>";
```
Comment (moozzyk) [inline samples/EchoApp/Startup.cs]: same as above

**Comment by moozzyk** [comment]
LGTM

**Comment by anurse** [comment]
The CoreFX code has been merged. Now merging this PR. Ping @Tratcher

**Pull request merged**